pypoptools/pypoptesting/test_framework.py

```diff
@@ -0,0 +1,182 @@
+            self.log.warning("Test Skipped: %s" % e.message)
+            self.success = TestStatus.SKIPPED
+        except AssertionError:
+            self.log.exception("Assertion failed")
+            self.success = TestStatus.FAILED
+        except KeyError:
+            self.log.exception("Key error")
+            self.success = TestStatus.FAILED
+        except subprocess.CalledProcessError as e:
+            self.log.exception("Called Process failed with '{}'".format(e.output))
+            self.success = TestStatus.FAILED
+        except Exception:
+            self.log.exception("Unexpected exception caught during testing")
+            self.success = TestStatus.FAILED
+        except KeyboardInterrupt:
+            self.log.warning("Exiting after keyboard interrupt")
+            self.success = TestStatus.FAILED
+        finally:
+            exit_code = self.shutdown()
+            sys.exit(exit_code)
+
+    def setup(self):
+        """Call this method to start up the test framework object with options set."""
+        # Set up temp directory and start logging
+        timestamp = datetime.datetime.now().strftime("%y%m%d_%H%M%S")
+        self.temp_dir = tempfile.mkdtemp(prefix=TMPDIR_PREFIX+timestamp)
+
+        self._start_logging()
+
+        self.skip_test_if_missing_module()
+        self.setup_network()
+
+        self.success = TestStatus.PASSED
+
+    def skip_test_if_missing_module(self):
+        """Override this method to skip a test if a module is not compiled"""
+        pass
+
+    def add_options(self, parser):
+        """Override this method to add command-line options to the test"""
+        pass
+
+    def set_test_params(self):
+        """Tests must this method to change default values for number of nodes, topology, etc"""
+        raise NotImplementedError
+
+    def run_test(self):
+        """Tests must override this method to define test logic"""
+        raise NotImplementedError
+
+    def setup_network(self):
+        """Tests must override this method to setup test network topology"""
+        raise NotImplementedError
+
+    def shutdown(self) -> int:
+        """Call this method to shut down the test framework object."""
+
+        if self.success == TestStatus.PASSED:
+            self.log.info("Tests successful")
+            exit_code = TEST_EXIT_PASSED
+        elif self.success == TestStatus.SKIPPED:
+            self.log.info("Test skipped")
+            exit_code = TEST_EXIT_SKIPPED
+        else:
+            self.log.error("Test failed. Test logging available at %s/test_framework.log", self.temp_dir)
+            exit_code = TEST_EXIT_FAILED
+
+        return exit_code
+
+    def stop_nodes(self):
+        """Stop multiple test nodes"""
+        for node in self.nodes:
+            # Issue RPC to stop nodes
+            node.stop()
+
+    def _create_nodes_(self, factory_lambda: Callable[[], Node]):
+        for i in range(self.num_nodes):
+            self.nodes.append(factory_lambda)
```

Comment:
here you make a list of factories, whereas you should call factory to get Node instance and then append it to this list

Reply:
so should be 

```
self.nodes.append(factory_lambda(dir))
```